File: coco_scala/src/main/scala/exp/m1/M1.scala
package exp.m1

trait Factory[T] extends exp.m0.Factory[T] {
  def sub(left: exp.Exp[T], right: exp.Exp[T]) : exp.Exp[T]
}

trait Sub[T] extends exp.m0.Exp[T] with Factory[T] {
  def left : exp.Exp[T]
  def right : exp.Exp[T]

  def eval: Double = left.eval - right.eval
}

object finalized {
  import exp.m0.finalized.Exp

  trait Factory extends exp.m1.Factory[Exp] with exp.m0.finalized.Factory {
    override def sub(left: exp.Exp[Exp], right: exp.Exp[Exp]) : Exp = new Sub(left, right)
  }

  class Sub(val left: exp.Exp[Exp], val right: exp.Exp[Exp]) extends exp.m1.Sub[Exp] with Exp with Factory
}

File: coco_csharp/CoCo.Tests/M1.cs
using System;
using Xunit;
using EP.M1;

namespace CoCo.Tests.M1
{
    public interface TestTemplate<FT> : CoCo.Tests.M0.TestTemplate<FT>, Factory<FT> {
        new public void test()
        {
            CoCo.Tests.M0.TestTemplate<FT> parent = this;
            parent.test();
            EP.Exp<FT> expr1 = sub(lit(1.0), lit(2.0));
            Assert.Equal(-1.0, convert(expr1).eval());

            EP.Exp<FT> expr2 = lit(2.0);
            Assert.Equal(2.0, convert(expr2).eval());

            Assert.Equal(-1.0, convert(sub(lit(1.0), lit(2.0))).eval());
        }
    }

    public class ActualTest : TestTemplate<EP.M0.Finalized.Exp>, EP.M1.Finalized.Factory {}

    public class M1Test
    {
        [Fact]
        public void TestM1()
        {
            TestTemplate<EP.M0.Finalized.Exp> theTest = new ActualTest();
            theTest.test();
        }
    }
}
